src/metabase/query_processor/middleware/optimize_temporal_filters.clj

```diff
@@ -0,0 +1,205 @@
+  the filter clause they're in?"
+  [field temporal-value]
+  (mbql.u/match-one temporal-value
+    [:relative-datetime (_ :guard #{0 :current})]
+    true
+
+    [(_ :guard #{:absolute-datetime :relative-datetime}) _ (unit :guard optimizable-units)]
+    (= (datetime-field-unit field) unit)))
+
+(defmethod can-optimize-filter? :default
+  [filter-clause]
+  (mbql.u/match-one filter-clause
+    [_
+     (field :guard optimizable-field?)
+     (temporal-value :guard optimizable-temporal-value?)]
+    (field-and-temporal-value-have-compatible-units? field temporal-value)))
+
+(defmethod can-optimize-filter? :between
+  [filter-clause]
+  (mbql.u/match-one filter-clause
+    [_
+     (field :guard optimizable-field?)
+     (temporal-value-1 :guard optimizable-temporal-value?)
+     (temporal-value-2 :guard optimizable-temporal-value?)]
+    (and (field-and-temporal-value-have-compatible-units? field temporal-value-1)
+         (field-and-temporal-value-have-compatible-units? field temporal-value-2))))
+
+(s/defn ^:private temporal-literal-lower-bound [unit t :- java.time.temporal.Temporal]
+  (:start (u.date/range t unit)))
+
+(s/defn ^:private temporal-literal-upper-bound [unit t :- java.time.temporal.Temporal]
+  (:end (u.date/range t unit)))
+
+(defn- change-datetime-field-unit-to-default [field]
+  (mbql.u/replace field
+    [:datetime-field wrapped _]
+    [:datetime-field wrapped :default]))
+
+(defmulti ^:private temporal-value-lower-bound
+  "Get a clause representing the *lower* bound that should be used when converting a `temporal-value-clause` (e.g.
+  `:absolute-datetime` or `:relative-datetime`) to an optimized range."
+  {:arglists '([temporal-value-clause datetime-field-unit])}
+  mbql.u/dispatch-by-clause-name-or-class)
+
+(defmulti ^:private temporal-value-upper-bound
+  "Get a clause representing the *upper* bound that should be used when converting a `temporal-value-clause` (e.g.
+  `:absolute-datetime` or `:relative-datetime`) to an optimized range."
+  {:arglists '([temporal-value-clause datetime-field-unit])}
+  mbql.u/dispatch-by-clause-name-or-class)
+
+(defmethod temporal-value-lower-bound :absolute-datetime
+  [[_ t unit] _]
+  [:absolute-datetime (temporal-literal-lower-bound unit t) :default])
+
+(defmethod temporal-value-upper-bound :absolute-datetime
+  [[_ t unit] _]
+  [:absolute-datetime (temporal-literal-upper-bound unit t) :default])
+
+(defmethod temporal-value-lower-bound :relative-datetime
+  [[_ n unit] datetime-field-unit]
+  [:relative-datetime (if (= n :current) 0 n) (or unit datetime-field-unit)])
+
+(defmethod temporal-value-upper-bound :relative-datetime
+  [[_ n unit] datetime-field-unit]
+  [:relative-datetime (inc (if (= n :current) 0 n)) (or unit datetime-field-unit)])
+
+(defmulti ^:private optimize-filter
+  "Optimize a filter clause agat a bucketed `:datetime-field` clause and `:absolute-datetime` or `:relative-datetime`
```

Comment:
"agat" again?  Just trying to understand in case it's a term I haven't heard.

Reply:
I think it was just a typo, I used that same sentence as the starting point for the docstring for both the namespace and that function